### ISSUE TITLE
support spaceship operator

### ISSUE THREAD

**Issue by aspiers** [supercollider/supercollider#3415]
Sorting can be supported more cleanly if there is a [Three-way comparison](https://en.wikipedia.org/wiki/Three-way_comparison) operator.  For example, given:

```
+ Object {
	<=> { |other|
		^case
		{ this == other } { nil  }
		{ this  < other } { true }
		{ false };
	}
}

TestObjectSpaceshipExtension : UnitTest {
	test_equality {
		this.assertEquals(1 <=> 1, nil, "equality should produce nil");
	}

	test_less_than {
		this.assertEquals(1 <=> 2, true, "less than should produce true");
	}

	test_greater_than {
		this.assertEquals(2 <=> 1, false, "greater than should produce false");
	}
}
```

we can easily sort based on futher criteria if the initial comparison
results in a tie, e.g.

    [ [ 3, 5 ], [ 2, 5 ], [ 6, 4 ] ].sort { |a, b|
        a[1] <=> b[1] ?? { a[0] <=> b[0] }
    }

yields

    [ [ 6, 4 ], [ 2, 5 ], [ 3, 5 ] ]

I'm not 100% sure but IIRC this would also allow stable sort implementations.

**Comment by telephon** [comment]
good idea!

In Ruby, the implementation distinguishes four cases, which seems correct:

http://ruby-doc.org/core-2.5.0/Object.html#method-i-3C-3D-3E

At least the `nil` case should probably include also those objects that can't be compared, that is where `<` fails.

**Comment by aspiers** [comment]
Yes, Ruby can do that because sorting functions are expected to typically return `0` or `-1` or `1` (with `nil` being an edge case).  However sclang currently expects the function provided to `sort` to return either `true` or `false`, so if we need to stick to this, the only obvious value left for a third state meaning "a and b are equal" is `nil`.  If sclang could be extended to alternatively accept sorting functions which return `0` / `-1` / `1` / `nil` then we could indeed copy Ruby's approach.

However, even if we did this, it would not make multi-criteria sorting any easier.  It's easy in Ruby because you can do:

    [[3, 5], [2, 5], [6, 4]].sortBy { |a| [a[1], a[0]] }

which relies on the fact that Ruby can sort nested arrays by default:

    [[5, 3], [5, 2], [4, 6]].sort   # => [[4, 6], [5, 2], [5, 3]]

but sclang can't (yet; maybe that's worth adding).

Also worth noting that even in the Ruby case where this works, this approach is not efficient when calculating the return value from the sorting function is expensive.  See https://en.wikibooks.org/wiki/Algorithm_Implementation/Sorting/Schwartzian_transform#Ruby for the Schwartzian Transform trick which tries to solve this problem, but actually seems to have further unsolved (and unstated!) problems:

- The mtime of every file is retrieved even though it might not be needed.
- It seems to incorrectly assume that `0 or 1` evaluates to `1`, but it does not. Therefore if two files tie on size, their age and name are ignored for the purposes of sorting.

Hrm, now I have to resurrect my Wikibooks account to fix that :-/

**Comment by aspiers** [comment]
Offtopic, but just for the curious - [I just asked on StackOverflow what would be the most elegant way of achieving this in Ruby](https://stackoverflow.com/questions/48240069/idiomatic-lazy-sorting-by-multiple-criteria).

**Comment by telephon** [comment]
if we had a compact way to map from `0, -1, 1, nil` to   Boolean , would this help ?

**Comment by aspiers** [comment]
Not sure exactly what you mean, but probably!

**Comment by telephon** [comment]
I don't have the time to look into it right now. But I like the idea that we get an optimal implementation in sclang for this.

**Comment by scztt** [comment]
Splitting operator `<` or `>` implementations into multiple places is a big risk - if equality operators are not logically equivalent, sort operations of that type of object can be undefined, and potentially even non-terminating. 3-way operator isn't a bad idea, as long as we used it as a full replacement for greater-than/less-than operators (where these just point to the 3-way). Parallel `>`, `<`, `<=>`, and `==` implementations is a disaster waiting to happen.

I'd be concerned about the performance impact as well. Currently, binary operator calls are implemented as a single instruction without sclang branching (I think something like `push value_a; push value_b; push operator_<; do_binary_op;` or whatever). Adding a level of indirection (2-3 additional sclang function calls) to all equality operators could have some serious overhead. Maybe worth mocking up some trivial cases of a `<=>` operator and seeing how much slower it is for sorting than calling `<` directly. 

(I realize that my "multiple implementations" fear is in direct conflict with my "too much indirection is slow" fear 😀  - the trials of proper strict-weak-ordering).

**Comment by LFSaw** [comment]
administrative: could we please have a more telling title for this issue? Something that tells about functionality rather than is a description of how the operator looks like...

(I am sorry that I could not come up with a title suggestion myself... :/ )
